### PR TITLE
release 0.6.0 – comprehensive ModelConfig refactor to support compatible HuggingFace dev/schnell models

### PR DESCRIPTION
**Pull request by anthonywu**
## change

1. a re-attempt of #84 but this time I don't have to explicitly endorse any org (i.e. Freepik or Shuttle) It should reduce expectation that the maintainers actually support the models, and eliminate future asks to support yet-another pre-config.
2. support issues like #102 (let users go ahead and test any HF model, as long as they bring the `--base-model` arg when needed
3. re-establish model config before going into #101 to support the Flux.1 Tools

## goals

1. keep the prior behavior as-is and pass existing tests and add backwards compatibility test cases
2. maintainers can continue to add the default configs list
3. early adopters can get readme/issues-level hints on how to use reportedly-compatible models

## demo: shuttle-3

https://huggingface.co/shuttleai/shuttle-3-diffusion

```sh
# didn't expect this to work, given shuttle-3 derives from schnell 
mflux-generate \
  --seed 42 \
  --base-model dev \
  --model shuttleai/shuttle-3-diffusion \
  --steps 4 \
  --prompt "A cat holding a sign that says hello world"
```
image

now, running as schnell, actually looks... better??

```sh
# same as above but with schnell
mflux-generate \
  --seed 42 \
  --base-model schnell \ 
  --model shuttleai/shuttle-3-diffusion \
  --steps 4 \
  --prompt "A cat holding a sign that says hello world"
```

image

### performance

- M1 Max 64GB, ~24s/step, 4 steps ~= 1m38s runtime

## demo: Freepik

https://huggingface.co/Freepik/flux.1-lite-8B-alpha

```sh
mflux-generate --base-model dev --steps 22 --model Freepik/flux.1-lite-8B-alpha --prompt "A cat holding a sign that says hello world"
```

image

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -254,7 +254,7 @@ Or, with the correct python environment active, create and run a separate script
 from mflux import Flux1, Config
 
 # Load the model
-flux = Flux1.from_alias(
```
Comment (anthonywu) [inline README.md:257]: I leave `from_alias` usable so there's no breakage, it's in the test cases

File: src/mflux/config/model_config.py [hunk 2]
```diff
@@ -1,27 +1,98 @@
-from enum import Enum
+from dataclasses import dataclass
+from typing import Literal
 
+DEFAULT_TRAIN_STEPS = 1000
 
-class ModelConfig(Enum):
-    FLUX1_DEV = ("black-forest-labs/FLUX.1-dev", "dev", 1000, 512)
-    FLUX1_SCHNELL = ("black-forest-labs/FLUX.1-schnell", "schnell", 1000, 256)
+KNOWN_SEQUENCE_LENGTH_BY_BASE_MODEL = {"dev": 512, "schnell": 256}
```
Comment (anthonywu) [inline src/mflux/config/model_config.py:6]: defining the magic numbers

File: src/mflux/config/model_config.py [hunk 3]
```diff
@@ -1,27 +1,98 @@
...
+    max_sequence_length: int
+    supports_guidance: bool
+    base_model: str | None
+
+    @property
+    def alias(self):
+        # maintain compatibility with < 0.4.0 behavior
+        # where alias is the name of an official model
+        if self.model_name.startswith("black-forest-labs/FLUX.1-"):
+            return self.model_name[len("black-forest-labs/FLUX.1-") :].lower()
```
Comment (anthonywu) [inline src/mflux/config/model_config.py:30]: other values to come with #101 will be `canny-dev`, `depth-dev` and `redux-dev` - the intention is to have the aliases support first party BFL models according to *their* naming convention

image

File: src/mflux/config/model_config.py [hunk 4]
```diff
@@ -1,27 +1,98 @@
...
+        supports_guidance=False,
+        base_model=None,
+    ),
+}
+
+
+class ModelConfig:
+    # keep these class members to be backwards compatible with < 0.5.0 ModelConfig Enum implementation
+    FLUX1_DEV = DefaultModelConfigs["dev"]
+    FLUX1_SCHNELL = DefaultModelConfigs["schnell"]
```
Comment (anthonywu) [inline src/mflux/config/model_config.py:55]: backward compatibility is maintained in the test suite

**Review by anthonywu — COMMENTED** [review]
Feels pretty good to me, I will make sure my image gen local tests pass.

Users who asked for #102 please help test!

File: src/mflux/config/model_config.py [hunk 5]
```diff
@@ -1,27 +1,98 @@
...
+        num_train_steps=DEFAULT_TRAIN_STEPS,
+        max_sequence_length=KNOWN_SEQUENCE_LENGTH_BY_BASE_MODEL["dev"],
+        supports_guidance=True,
+        base_model=None,
+    ),
+    "schnell": ModelAttrs(
+        model_name="black-forest-labs/FLUX.1-schnell",
+        num_train_steps=DEFAULT_TRAIN_STEPS,
+        max_sequence_length=KNOWN_SEQUENCE_LENGTH_BY_BASE_MODEL["schnell"],
+        supports_guidance=False,
```
Comment (anthonywu) [inline src/mflux/config/model_config.py:46]: ⚠️ turns out this broad assumption about guidance support may be invalid, the assumption is from the official schnell model docs: https://huggingface.co/black-forest-labs/FLUX.1-schnell

but shuttle-3-diffusion claims to support guidance even though it's based on schnell: https://huggingface.co/shuttleai/shuttle-3-diffusion

> Shuttle 3 Diffusion uses Flux.1 Schnell as its base.

and their docs allow guidance

image

**Comment by filipstrand** [comment]
@anthonywu Hi, sorry for my delayed response! I've been busy with the final things for the Dreambooth training feature so I have not had the time to look at you PR yet but will do that first thing when I have a PR up for that. Your results here look really good with the different models and nice that you could achieve this with a relatively small code change!

Dreabooth is pretty near completion and will be a big update (~2000 added lines now when I checked, but mostly new files in a speparate package, so hopefully not that many conflicts) . There are still more things I would like to improve, but I think it is good enough for a v1 release of this kind of feature. I was thinking of including that in the 0.5 release as well, but let's discuss that once I have a PR up for that feature, hopefully in a few days :)

**Comment by anthonywu** [comment]
@filipstrand Feel free to merge when you're satisfied with this PR and make edits as needed. I am traveling in the weeks ahead and have limited ability to make significant edits.

**Comment by filipstrand** [comment]
> @filipstrand Feel free to merge when you're satisfied with this PR and make edits as needed. I am traveling in the weeks ahead and have limited ability to make significant edits.

Alright, sounds good!

**Comment by filipstrand** [comment]
Quick update: I just merged the Dreambooth branch into main which unfortunately resulted in some conflicts with this branch, hopefully not too bad :)  

I'll publish a v.0.5.0 release that only includes the Dreambooth feature this time, but will prioritise looking at this PR after the holidays!

Happy Christmas!

**Comment by anthonywu** [comment]
@filipstrand rebased on latest `main` and passed the two modified tests (other tests in-flight)

```
tests/test_model_config.py::test_from_alias_function_redirect PASSED                                    [  4%]
tests/test_model_config.py::test_model_config_class_members_and_alias PASSED                            [  8%]
tests/test_model_config.py::test_bfl_dev PASSED                                                         [ 13%]
tests/test_model_config.py::test_bfl_schnell PASSED                                                     [ 17%]
tests/test_model_config.py::test_community_dev_implicit_base_model PASSED                               [ 21%]
tests/test_model_config.py::test_community_schnell_implicit_base_model PASSED                           [ 26%]
tests/test_model_config.py::test_community_dev_explicit_base_model PASSED                               [ 30%]
tests/test_model_config.py::test_community_schnell_explicit_base_model PASSED                           [ 34%]
tests/test_model_config.py::test_model_config_error PASSED                                              [ 39%]
tests/test_model_config.py::test_invalid_base_model_error PASSED                                        [ 43%]
tests/arg_parser/test_cli_argparser.py::test_model_path_requires_model_arg PASSED                       [ 47%]
tests/arg_parser/test_cli_argparser.py::test_model_arg_not_in_file PASSED                               [ 52%]
tests/arg_parser/test_cli_argparser.py::test_model_arg_in_file PASSED                                   [ 56%]
tests/arg_parser/test_cli_argparser.py::test_base_model_arg_in_file PASSED                              [ 60%]
tests/arg_parser/test_cli_argparser.py::test_prompt_arg PASSED                                          [ 65%]
tests/arg_parser/test_cli_argparser.py::test_guidance_arg PASSED                                        [ 69%]
tests/arg_parser/test_cli_argparser.py::test_quantize_arg PASSED                                        [ 73%]
tests/arg_parser/test_cli_argparser.py::test_seed_arg PASSED                                            [ 78%]
tests/arg_parser/test_cli_argparser.py::test_steps_arg PASSED                                           [ 82%]
tests/arg_parser/test_cli_argparser.py::test_lora_args PASSED                                           [ 86%]
tests/arg_parser/test_cli_argparser.py::test_image_to_image_args PASSED                                 [ 91%]
tests/arg_parser/test_cli_argparser.py::test_controlnet_args PASSED                                     [ 95%]
tests/arg_parser/test_cli_argparser.py::test_save_args PASSED                                           [100%]
```

**Comment by filipstrand** [comment]
@anthonywu This looks good and a great addition to the project! Will try this out today with `shuttle-3` and we can hopefully merge soon.  I have personally not played around much with other models, so great that you pushed for this to be implemented and it is really nice that  it requires relatively little change since the architecture are assumed to be the same as the Flux family.  On this topic I recently read about https://huggingface.co/ostris/Flex.1-alpha, and in that case it sounds like there are some slight variations in the number of `JointTransformerBlock` blocks (8 vs 19) in that case, so there could be reasons for us to have this a bit more flexible in the future to allow for slight variations in architectures for other models. 

In terms of testing for this PR, since we are not modifying the architecture in any way, I don't think it is necessary to also have image tests for other models besides the ones we already have.

File: src/mflux/config/model_config.py [hunk 6]
```diff
@@ -83,16 +85,14 @@ def from_name(
             max_sequence_length = KNOWN_SEQUENCE_LENGTH_BY_BASE_MODEL["schnell"]
 
         return ModelConfig(
-            alias,  # actually this arg is model_name
```
Comment (filipstrand) [inline src/mflux/config/model_config.py]: Added one extra commit here and removed the alias property on the `ModelConfig`... my thought was that we don't need this concept internally at all (we just use the model_name)?  If I have missed any special case of backwards compatibility with this commit, please let me know.

File: src/mflux/flux/flux.py [hunk 7]
```diff
@@ -135,16 +136,22 @@ def generate_image(
             config=config,
         )
 
+    @staticmethod
+    def from_alias(alias: str, quantize: int | None = None) -> "Flux1":
+        warnings.warn(
+            "from_alias is deprecated and will be removed in a future release. " "Please use from_name instead.",
```
Comment (anthonywu) [inline src/mflux/flux/flux.py:142]: the `ruff format` action altered this string strangely, should edit to just one set of quotes

File: src/mflux/config/model_config.py (hunk 6 above)
Comment (anthonywu) [reply src/mflux/config/model_config.py]: personally I prefer `model_name` but I was operating under a strict interpretation of semantic versioning and wanted to respect what came before, and any devs that might have sub-classed `mflux` objects and hard coded the `alias=` kwarg.

I am happy with the deprecation and given this project should be considered "beta" status, I don't think this change will offend anyone.

**Review by anthonywu — COMMENTED** [review]
One remaining concern of mine: the example generations I provided in the original summary now generates slightly different images. I am wondering if this is a change in `mflux`, `mlx` (precedent here), or the model files have been updated ever slightly.

Given we've not released this yet, and there's no baseline to compare to, I think it's okay to release and let the users have the feature then act on bug reports if there's some missed expectations.

File: src/mflux/flux/flux.py (hunk 7 above)
Comment (filipstrand) [reply src/mflux/flux/flux.py:142]: Fixed!

File: src/mflux/config/model_config.py (hunk 6 above)
Comment (filipstrand) [reply src/mflux/config/model_config.py]: True, that could happen. I agree, it should probably be pretty safe to update this as you say. 

I'll try to remember to document this change in the release-notes for v.0.6.0.

**Comment by filipstrand** [comment]
> One remaining concern of mine: the example generations I provided in the original summary now generates slightly different images. I am wondering if this is a change in `mflux`, `mlx` (precedent here), or the model files have been updated ever slightly.
> 
> Given we've not released this yet, and there's no baseline to compare to, I think it's okay to release and let the users have the feature then act on bug reports if there's some missed expectations.

This is due to my latest updates in https://github.com/filipstrand/flux/pull/116. After that PR I also had to update my test images a bit (there were some visible changes in the final images, but things you would mostly see if you compared them side by side). In that PR I did some refactoring recently involving the attention operation (making use of `scaled_dot_product_attention` in `from mlx.core.fast`) For this I had to rewrite some parts to use the provided `mask` argument in `CLIPSdpaAttention`. While gaining no real speedup on my M1 Pro machine, it seemed a bit cleaner and this change resulted a very minor change in image output. 

However, since mlx v.0.22.0 I had to cast the precision of the `mask` to the appropriate level (`bfloat16`) (https://github.com/filipstrand/mflux/commit/16a0225b41959dd69f882039912d12cab67ccd92) for things to work at all (before v.0.22.0, that cast was not needed and was typically in `float32`), and that caused a visible change in the output images.  

If I reset to mlx==0.21.1 and remove the line 

```
causal_attention_mask = causal_attention_mask.astype(Config.precision)
```

 then I can recreate your image from before:
 

```
mflux-generate \
  --seed 42 \
  --base-model schnell \ 
  --model shuttleai/shuttle-3-diffusion \
  --steps 4 \
  --prompt "A cat holding a sign that says hello world"
```

image_5

But, as you say, I guess this change is fine.

**Comment by anthonywu** [comment]
Excellent documentation on the cause of the subtle difference! I feel satisfied enough to release to interest users and to close the issue.

**Comment by filipstrand** [comment]
Great! Then lets go ahead and merge this and I'll write up a little bit of documentation tomorrow before creating the release.
